### ISSUE TITLE
[CUSTOMER] Ping not working for non root users on Diskless PowerKVM guests (rhels7.2)

### ISSUE THREAD

**Issue by mew2057** [xcat2/xcat-core#922]
Hi, 
I've been playing with provisioning Diskless VM guests on PowerKVM and I ran into a bit of a problem. When I ssh into the guest and attempt to ping as a user my user is denied access to the ping command. This issue is resolved when I change permissions for /bin/ping, but it seems to represent an underlying permissions issue that I can't quite source. As far as I can tell SELinux is disabled on both the xcat master and guest nodes.

Thanks,
John

> [root@localhost ~]# ping 10.4.4.24
> PING 10.4.4.24 (10.4.4.24) 56(84) bytes of data.
> 64 bytes from 10.4.4.24: icmp_seq=1 ttl=64 time=0.152 ms
> 64 bytes from 10.4.4.24: icmp_seq=2 ttl=64 time=0.106 ms
> 64 bytes from 10.4.4.24: icmp_seq=3 ttl=64 time=0.060 ms
> ^C
> --- 10.4.4.24 ping statistics ---
> 3 packets transmitted, 3 received, 0% packet loss, time 1998ms
> rtt min/avg/max/mdev = 0.060/0.106/0.152/0.037 ms
>[root@localhost ~]# su - jdunham
> [jdunham@localhost ~]# ping 10.4.4.24
> ping: icmp open socket: Operation not permitted
> [jdunham@localhost ~]# exit
> [root@localhost ~]# chmod u+s /bin/ping
> [root@localhost ~]# su - jdunham
> Last login: Wed Apr 13 14:58:49 UTC 2016 on pts/0
> [jdunham@localhost ~]# ping 10.4.4.24
> PING 10.4.4.24 (10.4.4.24) 56(84) bytes of data.
> 64 bytes from 10.4.4.24: icmp_seq=1 ttl=64 time=0.447 ms
> 64 bytes from 10.4.4.24: icmp_seq=2 ttl=64 time=0.097 ms
> ^C
> --- 10.4.4.24 ping statistics ---
> 2 packets transmitted, 2 received, 0% packet loss, time 999ms
> rtt min/avg/max/mdev = 0.097/0.272/0.447/0.175 ms

**Comment by whowutwut** [comment]
@mew2057  Just doing some debug here... 

On rhels7.2 ppc64le, I do see this problem.  

sles12 ppc64le, it does not exist.. 
```
fs2vm112:~ # ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.128 ms
^C
--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.128/0.128/0.128/0.000 ms
fs2vm112:~ # useradd -m vhu
fs2vm112:~ # su - vhu
vhu@fs2vm112:~> ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.022 ms
^C
--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.022/0.022/0.022/0.000 ms
vhu@fs2vm112:~> ls -ltr /bin/ping
lrwxrwxrwx 1 root root 13 Apr 18 14:09 /bin/ping -> /usr/bin/ping

vhu@fs2vm112:~> cat /etc/*release*
SUSE Linux Enterprise Server 12 (ppc64le)
VERSION = 12
PATCHLEVEL = 0
# This file is deprecated and will be removed in a future service pack or release.
# Please check /etc/os-release for details about this release.
NAME="SLES"
VERSION="12"
VERSION_ID="12"
PRETTY_NAME="SUSE Linux Enterprise Server 12"
ID="sles"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:12"
```

rhels6.6 ppc64 , it doesn't exist either... 
```
[root@c910f02c05p06 ~]# uname -a
Linux c910f02c05p06 2.6.32-504.el6.ppc64 #1 SMP Tue Sep 16 01:52:45 EDT 2014 ppc64 ppc64 ppc64 GNU/Linux
[root@c910f02c05p06 ~]# cat /etc/*release*
Red Hat Enterprise Linux Server release 6.6 (Santiago)
Red Hat Enterprise Linux Server release 6.6 (Santiago)
cpe:/o:redhat:enterprise_linux:6server:ga:server

[root@c910f02c05p06 ~]# ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.032 ms
^C
--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 988ms
rtt min/avg/max/mdev = 0.032/0.032/0.032/0.000 ms
[root@c910f02c05p06 ~]# useradd -m vhu
[root@c910f02c05p06 ~]# su - vhu
[vhu@c910f02c05p06 ~]$ ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.022 ms
^C
--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 778ms
rtt min/avg/max/mdev = 0.022/0.022/0.022/0.000 ms
[vhu@c910f02c05p06 ~]$ ls -ltr /bin/ping
-rwsr-xr-x 1 root root 50152 Sep 17  2013 /bin/ping
```

let me see what else i can find....

**Comment by immarvin** [comment]
hi @whowutwut , I looked into the problem on your env with @neo954 today.

This problem is caused by the fact that the "capacity" attribute of "ping" binary is lost during cpio in packimage,  For the "capacity" of file, see https://www.insecure.ws/linux/getcap_setcap.html#examples for details.

This can be provided the following process:

````
[root@fs2vm111 ~]# su - vhu sh  -c "ping baidu.com"
ping: icmp open socket: Operation not permitted
[root@fs2vm111 ~]# setcap cap_net_raw+ep /usr/bin/ping ;su - vhu sh  -c "ping baidu.com"
PING baidu.com (180.149.132.47) 56(84) bytes of data.
64 bytes from 180.149.132.47: icmp_seq=1 ttl=42 time=412 ms
64 bytes from 180.149.132.47: icmp_seq=2 ttl=42 time=383 ms
...
[root@fs2vm111 ~]# setcap cap_net_raw-ep /usr/bin/ping ;su - vhu sh  -c "ping baidu.com"
ping: icmp open socket: Operation not permitted


[root@fs2vm100 ~]# getcap /install/netboot/rhels7.2/ppc64le/compute/rootimg/usr/bin/ping
/install/netboot/rhels7.2/ppc64le/compute/rootimg/usr/bin/ping = cap_net_admin,cap_net_raw+p
[root@fs2vm100 rootimg]# mv /install/netboot/rhels7.2/ppc64le/compute/rootimg.gz /tmp/rootimg/
[root@fs2vm100 rootimg]# cd  /tmp/rootimg/
[root@fs2vm100 rootimg]# gunzip -c ./rootimg.gz |cpio -idum
[root@fs2vm100 rootimg]# getcap ./usr/bin/ping
[root@fs2vm100 rootimg]#
````
cpio provided no option to preserve the  "capacity" , tar provides the following options on this:

````

       --xattrs
              Enable extended attributes support

       --xattrs-exclude=MASK
              specify the exclude pattern for xattr keys

       --xattrs-include=MASK
              specify the include pattern for xattr keys
````
A PR https://github.com/xcat2/xcat-core/issues from @jjohnson42 provides a way to fix this problem. however, it is an incomplete commit and has some problem if the osimage is packed with `` packimage -m txc rhels7.2-ppc64le-netboot-compute``, I will try to fix the problems of the PR in another PR, another problem for "-m txc" is that it is very very slow and spent a lot of time(about 10+ minutes for rootimg around 200M)  and memory, I am not sure whether it is acceptable.

Back to this problem, I think we need to provide a  option to use ``tar`` instead of ``cpio`` in ``packimage`` or replace ``cpio`` with ``tar``.

**Comment by immarvin** [comment]
ping @daniceexi

**Comment by mew2057** [comment]
I built the most recent version of xCAT to try the `packimage -m txc`, and I find that it doesn't produce anything and I get this error:

> Error: No packed image for platform rhels7.2, architecture ppc64le, and profile compute found at /install/netboot/rhels7.2/ppc64le/compute-dev-02/rootimg.gz or /install/netboot/rhels7.2/ppc64le/compute-dev-02/rootimg.sfs on c931f04p26-mgt, please run packimage (e.g.  packimage -o rhels7.2 -p compute -a ppc64le

Thoughts on a work around? The issue is blocking.

**Comment by immarvin** [comment]
hi @mew2057 , the team need some time to finish the code and verification of preserving the capabilities of files in "packimage" , might need to replace "cpio" with "tar".

As a quick workaround, run the following script(can be named "gencapscript") on your management node, such as ``gencapscript rhels7.2-ppc64le-netboot-compute``
````
#!/bin/bash
[ -z "$1" ] && echo "USAGE:  gencapscript <diskless osimage name>" && exit 1
PWD=`pwd`

#get the rootimg path of the specified diskless osimage
ROOTIMGDIR=`lsdef -c -t osimage -o $1 -i rootimgdir|awk -F "rootimgdir=" '{print $2}'`
[ -z "ROOTIMGDIR" ] && echo "unable to get rootimgdir of osimage $1" && exit 1
cd $ROOTIMGDIR/rootimg/
#generate the postscript "restorecap"
#which includes the commands to restore the capablity of the files on diskless node
{
echo "#!/bin/bash" 
echo "cd /"
find . |xargs getcap|awk -F= '{print "setcap" $2 " " $1}'
} > /install/postscripts/restorecap
chmod +x /install/postscripts/restorecap
#set "restorecap" as the postbootscript of the specified osimage
chdef -t osimage -o $1 -p postbootscripts=restorecap
cd $PWD
````
 then run ``nodeset``... ``rpower``... to reprovision the diskless node

**Comment by mew2057** [comment]
gencapscript functions as an adequate workaround.

**Comment by whowutwut** [comment]
@tingtli  I marked this as "testcase requested".  You can associate a test case pull request with this issue and then change the label to "testcase_added" .  I would like to add a testcase to verify that the OS provisioned is OK.  (For both diskful and diskless, if possible.

We can create a new testing suite that does "OS verification" and could add more scenarios in the future.  For the very basic test, we should replicate this ping test as root and also as non root so we can catch this problem.  

```
ping -c 1 127.0.0.1 
useradd -m xcatuser 
su - xcatuser sh  -c "ping -c 1 127.0.0.1"
```
Check the RC for both pings.

**Comment by whowutwut** [comment]
@immarvin  I set this to 2.11.2 since this is the branch that @mew2057  is using at this time.  i'll add this into the known issues for 2.11.1 release

**Comment by daniceexi** [comment]
The fix for this issue will be big, so move the fix to 2.12.1
